File: go/problems/p1.go
package problems

func twoSum(nums []int, target int) []int {
	numIndices := make(map[int][]int)
	for i, n := range nums {
		numIndices[n] = append(numIndices[n], i)
	}

	for m, indices := range numIndices {
		n := target - m

		keyInMap := false
		for k := range numIndices {
			if k == n {
				keyInMap = true
			}
		}
		if keyInMap {
			if m == n {
				if 2 <= len(indices) {
					return indices[:2]
				}
			} else {
				return []int{indices[0], numIndices[n][0]}
			}
		}
	}

	return nil
}

File: python/leetcode/problems/p1.py
from typing import List
from collections import defaultdict


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        num_indices = defaultdict(list)
        for i, n in enumerate(nums):
            num_indices[n].append(i)
        for m, indices in num_indices.items():
            n = target - m
            if n in num_indices:
                if n == m:
                    if 2 <= len(indices):
                        return indices[:2]
                else:
                    return [indices[0], num_indices[n][0]]
